opa/pkg/main.go

```diff
@@ -0,0 +1,324 @@
+					})
+					logrus.Warnf("Error while retrieving API Version %+v", err)
+					return actionItems, nil
+				}
+				gvr := mapping.Resource
+				list, err := dynamicInterface.Resource(gvr).Namespace("").List(ctx, metav1.ListOptions{})
+				if err != nil {
+					return nil, err
+				}
+				for _, obj := range list.Items {
+					query, err := rego.New(
+						rego.Query("results = data"),
+						rego.Module("fairwinds", check.Spec.Rego),
+						rego.Function1(
+							&rego.Function{
+								Name: "kubernetes",
+								Decl: types.NewFunction(types.Args(types.S), types.S),
+							},
+							getKubernetesData),
+					).PrepareForEval(ctx)
+					if err != nil {
+						actionItems = append(actionItems, ActionItem{
+							Title:       checkInstance.Name,
+							Description: "An error occured parsing the Rego for this custom check",
+							Remediation: "Make sure that the Rego is valid for this check.",
+							Severity:    0.4,
+							Category:    "Reliability",
+						})
+						logrus.Warnf("Error while parsing Rego %+v", err)
+						return actionItems, nil
+					}
+					obj.Object["parameters"] = checkInstance.Spec.Parameters
+					// TODO Find another way to get parameters in - Should they be a function or input?
+					// TODO Caching
+					evaluatedInput := rego.EvalInput(obj.Object)
+					results, err := query.Eval(ctx, evaluatedInput)
+					if err != nil {
+						return nil, err
+					}
+					newItems, err := processResults(obj, results, check, checkInstance)
+					if err != nil {
+						actionItems = append(actionItems, ActionItem{
+							Title:       checkInstance.Name,
+							Description: "An error occured processing the results of this check.",
+							Remediation: "Make sure that the return values are all correct.",
+							Severity:    0.4,
+							Category:    "Reliability",
+						})
+						logrus.Warnf("Error while parsing results %+v", err)
+						return actionItems, nil
+					}
+					actionItems = append(actionItems, newItems...)
+				}
+			}
+		}
+	}
+	return actionItems, nil
+}
+
+func getOutputArray(results rego.ResultSet) []interface{} {
+	returnSet := make([]interface{}, 0)
+
+	for _, result := range results {
+		for _, pack := range result.Bindings["results"].(map[string]interface{}) {
+			for _, outputArray := range pack.(map[string]interface{}) {
+				for _, output := range outputArray.([]interface{}) {
+					returnSet = append(returnSet, output)
+				}
+			}
+		}
+	}
+	return returnSet
+}
+
+func processResults(resource unstructured.Unstructured, results rego.ResultSet, check customCheck, checkInstance customCheckInstance) ([]ActionItem, error) {
+	actionItems := make([]ActionItem, 0)
+	instanceOutput := checkInstance.Spec.Output
+	checkOutput := check.Spec.Output
+	for _, output := range getOutputArray(results) {
+		severity := checkOutput.Severity
+		title := checkOutput.Title
+		remediation := checkOutput.Remediation
+		category := checkOutput.Category
+		if instanceOutput.Severity != nil {
```

Comment:
For my understanding:
* `checkOutput` is the check definition
* `instanceOutput` is the check instance (which links the check to certain resources)
* `output` is what came out of running the rego

correct?

Reply:
Yes